src/ripple/core/Pg.cpp

```diff
@@ -0,0 +1,1420 @@
+        }
+
+        if (!option->val ||
+            (remember_ip &&
+             (!strcmp(option->keyword, "hostaddr") ||
+              !strcmp(option->keyword, "port"))))
+        {
+            continue;
+        }
+
+        if (strlen(option->keyword) > maxFieldSize ||
+            strlen(option->val) > maxFieldSize)
+        {
+            std::stringstream ss;
+            ss << "DB returned a connection option name or value with\n";
+            ss << "excessive size (>" << maxFieldSize << " bytes).\n";
+            ss << "option (possibly truncated): "
+               << std::string_view(
+                      option->keyword,
+                      std::min(strlen(option->keyword), maxFieldSize))
+               << '\n';
+            ss << " value (possibly truncated): "
+               << std::string_view(
+                      option->val, std::min(strlen(option->val), maxFieldSize));
+            Throw<std::runtime_error>(ss.str());
+        }
+        config_.keywords.push_back(option->keyword);
+        config_.values.push_back(option->val);
+    }
+
+    config_.keywordsIdx.reserve(config_.keywords.size() + 1);
+    config_.valuesIdx.reserve(config_.values.size() + 1);
+    for (std::size_t n = 0; n < config_.keywords.size(); ++n)
+    {
+        config_.keywordsIdx.push_back(config_.keywords[n].c_str());
+        config_.valuesIdx.push_back(config_.values[n].c_str());
+    }
+    config_.keywordsIdx.push_back(nullptr);
+    config_.valuesIdx.push_back(nullptr);
+
+    get_if_exists(pgConfig, "max_connections", config_.max_connections);
+    std::size_t timeout;
+    if (get_if_exists(pgConfig, "timeout", timeout))
+        config_.timeout = std::chrono::seconds(timeout);
+}
+
+void
+PgPool::setup()
+{
+    {
+        std::stringstream ss;
+        ss << "max_connections: " << config_.max_connections << ", "
+           << "timeout: " << config_.timeout.count() << ", "
+           << "connection params: ";
+        bool first = true;
+        for (std::size_t i = 0; i < config_.keywords.size(); ++i)
+        {
+            if (first)
+                first = false;
+            else
+                ss << ", ";
+            ss << config_.keywords[i] << ": "
+               << (config_.keywords[i] == "password" ? "*" : config_.values[i]);
+        }
+        JLOG(j_.debug()) << ss.str();
+    }
+}
+
+void
+PgPool::onStop()
+{
+    std::lock_guard<std::mutex> lock(mutex_);
+    stop_ = true;
+    cond_.notify_all();
+    idle_.clear();
+    stopped();
+    JLOG(j_.info()) << "stopped";
+}
+
+void
+PgPool::idleSweeper()
+{
+    std::size_t before, after;
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
+        before = idle_.size();
+        if (config_.timeout != std::chrono::seconds(0))
+        {
+            auto const found =
+                idle_.upper_bound(clock_type::now() - config_.timeout);
+            for (auto it = idle_.begin(); it != found; ++it)
+            {
+                idle_.erase(it);
```

Comment:
@natenichols is seeing a segfault here on Mac, and I think it's because `erase()` invalidates the iterator. I think the way this should be done is:
```
            for (auto it = idle_.begin(); it != found;)
            {
                it = idle_.erase(it);
                --connections_;
            }
```
Or something analagous with a while loop. Here is a source that talks about this situation, though it's referring to std::map: https://stackoverflow.com/questions/4636182/problem-with-stdmapiterator-after-calling-erase

Reply:
That fixed it, thanks CJ.

Reply:
Here's the fix: https://github.com/mtrippled/rippled/commit/ae4a09f0891235ccfe05913389413d74c1fbcc7e